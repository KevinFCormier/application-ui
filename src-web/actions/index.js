/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/*
 * action types
 *
 * action name: <NOUN>_<VERB>
 */

export const CATALOG_DROPDOWN_FILTERS_VISIBILITY_TOGGLE = 'CATALOG_DROPDOWN_FILTER_VISIBILITY_TOGGLE'
export const CATALOG_FETCH_ERROR_STATUS_CHANGE = 'CATALOG_FETCH_ERROR_STATUS_CHANGE'
export const CATALOG_INSTALL_FAILURE = 'CATALOG_INSTALL_FAILURE'
export const CATALOG_INSTALL_LOADING = 'CATALOG_INSTALL_LOADING'
export const CATALOG_INSTALL_SUCCESS = 'CATALOG_INSTALL_SUCCESS'
export const CATALOG_INSTALL_VALIDATION_FAILURE = 'CATALOG_INSTALL_VALIDATION_FAILURE'
export const CATALOG_RESOURCE_FILTER_REPOS = 'CATALOG_RESOURCE_FILTER_REPOS'
export const CATALOG_RESOURCE_FILTER_SEARCH = 'CATALOG_RESOURCE_FILTER_SEARCH'
export const CATALOG_RESOURCE_SELECT = 'CATALOG_RESOURCE_SELECT' // clicking on a tile

export const CLEAR_REQUEST_STATUS = 'CLEAR_REQUEST_STATUS'

export const CREATE_GRAFANA_DASHBOARD = 'CREATE_GRAFANA_DASHBOARD'

export const DEL_RECEIVE_FAILURE = 'DEL_RECEIVE_FAILURE'
export const DEL_RECEIVE_SUCCESS = 'DEL_RECEIVE_SUCCESS'
export const DEL_REQUEST = 'DEL_REQUEST'

export const LOGS_RESET= 'LOGS_RESET'
export const LOGS_SEARCH = 'LOGS_SEARCH'

export const MODAL_UPDATE = 'MODAL_UPDATE'

export const NAMESPACES_RECEIVE_FAILURE = 'NAMESPACES_RECEIVE_FAILURE'
export const NAMESPACES_RECEIVE_SUCCESS = 'NAMESPACES_RECEIVE_SUCCESS'

export const NAV_MODIFY = 'NAV_MODIFY'
export const NAV_RECEIVE_SUCCESS = 'NAV_RECEIVE_SUCCESS'

export const POST_RECEIVE_FAILURE = 'POST_RECEIVE_FAILURE'
export const POST_RECEIVE_SUCCESS = 'POST_RECEIVE_SUCCESS'
export const POST_REQUEST = 'POST_REQUEST'
export const PUT_RECEIVE_FAILURE = 'PUT_RECEIVE_FAILURE'
export const PUT_RECEIVE_SUCCESS = 'PUT_RECEIVE_SUCCESS'
export const PUT_REQUEST = 'PUT_REQUEST'

export const REPO_FETCH_REQUEST_SUCCESS = 'REPO_FETCH_REQUEST_SUCCESS'

export const RESOURCES_FETCH_REQUEST_LOADING = 'RESOURCE_FETCH_REQUEST_LOADING'
export const RESOURCES_FETCH_REQUEST_SUCCESS = 'RESOURCES_FETCH_REQUEST_SUCCESS'
export const RESOURCE_ADD = 'RESOURCE_ADD'
export const RESOURCE_DELETE= 'RESOURCE_DELETE'
export const RESOURCE_MODIFY = 'RESOURCE_MODIFY'
export const RESOURCE_RECEIVE_FAILURE = 'RESOURCE_RECEIVE_FAILURE'
export const RESOURCE_RECEIVE_SUCCESS = 'RESOURCE_RECEIVE_SUCCESS'
export const RESOURCE_REQUEST = 'RESOURCE_REQUEST'
export const RESOURCE_RESET= 'RESOURCE_RESET'


export const DASHBOARD_RECEIVE_SUCCESS = 'DASHBOARD_RECEIVE_SUCCESS'
export const DASHBOARD_RECEIVE_FAILURE = 'DASHBOARD_RECEIVE_FAILURE'


export const ROLE_RECEIVE_FAILURE = 'ROLE_RECEIVE_FAILURE'
export const ROLE_RECEIVE_SUCCESS = 'ROLE_RECEIVE_SUCCESS'

export const SECONDARY_HEADER_UPDATE = 'SECONDARY_HEADER_UPDATE'

export const SWITCH_NAMESPACE = 'SWITCH_NAMESPACE'

export const TABLE_PAGE_CHANGE = 'TABLE_PAGE_CHANGE'
export const TABLE_SEARCH = 'TABLE_SEARCH'
export const TABLE_SORT = 'TABLE_SORT'

export const TOPOLOGY_FILTERS_UPDATE = 'TOPOLOGY_FILTERS_UPDATE'
export const TOPOLOGY_FILTERS_REQUEST = 'TOPOLOGY_FILTERS_REQUEST'
export const TOPOLOGY_FILTERS_RECEIVE_ERROR = 'TOPOLOGY_FILTERS_RECEIVE_ERROR'
export const TOPOLOGY_FILTERS_RECEIVE_SUCCESS = 'TOPOLOGY_FILTERS_RECEIVE_SUCCESS'
export const TOPOLOGY_SELECTION_UPDATE = 'TOPOLOGY_SELECTION_UPDATE'

export const UICONFIG_RECEIVE_SUCCESS = 'UICONFIG_RECEIVE_SUCCESS'

export const UPDATE_USERPREFERENCES = 'UPDATE_USERPREFERENCES'

export const USERPREFERENCES_RECEIVE_FAILURE = 'USERPRFERENCES_RECEIVE_FAILURE'
export const USERPREFERENCES_RECEIVE_SUCCESS = 'USERPRFERENCES_RECEIVE_SUCCESS'

export const USER_LOGIN_RECEIVE_SUCCESS = 'USER_LOGIN_RECEIVE_SUCCESS'
export const USER_LOGOUT_RECEIVE_FAILURE = 'USER_LOGOUT_RECEIVE_FAILURE'
export const USER_LOGOUT_RECEIVE_SUCCESS = 'USER_LOGOUT_RECEIVE_SUCCESS'
export const USER_LOGOUT_REQUEST = 'USER_LOGOUT_REQUEST'
/*
 * other constants specific to particular actions
 */

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE'
}

export const USER_LOGIN_STATUS = {
  LOGGED_IN: 'LOGGED_IN',
  LOGGED_OUT: 'LOGGED_OUT'
}

export const REQUEST_STATUS = {
  INCEPTION: 'INCEPTION',
  DONE: 'DONE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR'
}

export const PAGE_SIZES = {
  DEFAULT: 20,
  VALUES: [5, 10, 20, 50, 75, 100]
}

export const SORT_DIRECTION_ASCENDING = 'asc'
export const SORT_DIRECTION_DESCENDING = 'desc'
