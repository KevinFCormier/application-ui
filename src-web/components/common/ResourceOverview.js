/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import React from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Loading } from 'carbon-components-react';
import { connect } from 'react-redux';
import CountsCardModule from '../CountsCardModule';
import ChannelsCardModule from '../ChannelsCardModule';
import StructuredListModule from '../../components/common/StructuredListModule';
import {
  getSingleResourceItem,
  resourceItemByName,
} from '../../reducers/common';
import {
  getNumDeployables,
  getNumDeployments,
  getNumFailedDeployments,
} from '../../../lib/client/resource-helper';
import { withLocale } from '../../providers/LocaleProvider';
import resources from '../../../lib/shared/resources';
import msgs from '../../../nls/platform.properties';

resources(() => {
  require('../../../scss/resource-overview.scss');
});

const ResourceOverview = withLocale(({ staticResourceData, item, params, modules, resourceType, showAppDetails, locale }) => {
  if (!item) { return <Loading withOverlay={false} className="content-spinner" />; }
  const modulesRight = [];
  const modulesBottom = [];
  React.Children.map(modules, (module) => {
    if (module.props.right) {
      modulesRight.push(React.cloneElement(module, {
        staticResourceData,
        resourceType,
        resourceData: item,
        params,
      }));
    } else {
      modulesBottom.push(React.cloneElement(module, {
        staticResourceData,
        resourceType,
        resourceData: item,
        params,
      }));
    }
  });

  const countsCardData = [
    {
      msgKey: 'table.header.deployables',
      count: getNumDeployables(item),
    },
    {
      msgKey: 'table.header.deployments',
      count: getNumDeployments(item),
    },
    {
      msgKey: 'table.header.failedDeployments',
      count: getNumFailedDeployments(item),
    },
  ];

  const channelsCardData = [
    {
      name: 'Development',
      counts: {
        pending: {
          total: 3,
        },
        'in progress': {
          total: 2,
        },
        failed: {
          total: 1,
        },
      },
    },
    {
      name: 'QA',
      counts: {
        pending: {
          total: 3,
        },
        'in progress': {
          total: 2,
        },
        failed: {
          total: 1,
        },
      },
    },
    {
      name: 'Dev',
      counts: {
        pending: {
          total: 3,
        },
        'in progress': {
          total: 2,
        },
        failed: {
          total: 1,
        },
      },
    },
  ];

  return (
    <div className="overview-content">
      {!showAppDetails &&
        <React.Fragment>
          <div className="overview-content-bottom overview-content-with-padding">
            <CountsCardModule data={countsCardData} />
          </div>
          <div className="deployment-channels-title">
            {msgs.get('application.deployments.channels', locale)}
            {Array.isArray(channelsCardData) &&
              <span>&nbsp;({channelsCardData.length})</span>
            }
          </div>
          <div className="overview-content-bottom">
            <ChannelsCardModule data={channelsCardData} />
          </div>
        </React.Fragment>
      }
      {showAppDetails &&
        <StructuredListModule
          title={staticResourceData.detailKeys.title}
          headerRows={staticResourceData.detailKeys.headerRows}
          rows={staticResourceData.detailKeys.rows}
          data={item}
        />
      }
      {showAppDetails && modulesRight.length > 0 && (
        <div className="overview-content-right">{modulesRight}</div>
      )}
      {showAppDetails &&
        <div className="overview-content-bottom">{modulesBottom}</div>
      }
    </div>
  );
});

ResourceOverview.contextTypes = {
  locale: PropTypes.string,
};

ResourceOverview.propTypes = {
  item: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  modules: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  params: PropTypes.object,
  resourceType: PropTypes.object,
  staticResourceData: PropTypes.object,
};

const mapStateToProps = (state, ownProps) => {
  const { resourceType, params } = ownProps;
  const name = decodeURIComponent(params.name);
  const item = getSingleResourceItem(state, {
    storeRoot: resourceType.list,
    resourceType,
    name,
    predicate: resourceItemByName,
    namespace: params.namespace ? decodeURIComponent(params.namespace) : null,
  });
  return { item };
};

export default withRouter(connect(mapStateToProps)(withLocale(ResourceOverview)));
