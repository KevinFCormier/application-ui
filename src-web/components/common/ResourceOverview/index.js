/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import React from 'react'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { Loading } from 'carbon-components-react'
import { connect } from 'react-redux'
import CountsCardModule from '../../CountsCardModule'
import ChannelsCardCarousel from '../../ChannelsCardCarousel'
import ApplicationTopologyModule from '../../ApplicationTopologyModule'
import StructuredListModule from '../../../components/common/StructuredListModule'
import {
  getSingleResourceItem,
  resourceItemByName
} from '../../../reducers/common'
import {
  getChannelsList,
  getNumDeployables,
  getNumDeployments,
  getNumPendingDeployments,
  getNumInProgressDeployments,
  getNumFailedDeployments
} from './utils'
import { withLocale } from '../../../providers/LocaleProvider'
import resources from '../../../../lib/shared/resources'
import InstancesTable from '../../InstancesTable'

resources(() => {
  require('./style.scss')
})

const ResourceOverview = withLocale(
  ({
    staticResourceData,
    item,
    channelList,
    params,
    modules,
    resourceType,
    actions,
    showAppDetails,
    showExpandedTopology,
    getVisibleResources,
    incidentCount
  }) => {
    if (!item) {
      return <Loading withOverlay={false} className="content-spinner" />
    }
    const modulesRight = []
    const modulesBottom = []
    React.Children.map(modules, module => {
      if (module.props.right) {
        modulesRight.push(
          React.cloneElement(module, {
            staticResourceData,
            resourceType,
            resourceData: item,
            params
          })
        )
      } else {
        modulesBottom.push(
          React.cloneElement(module, {
            staticResourceData,
            resourceType,
            resourceData: item,
            params
          })
        )
      }
    })
    const countsCardData = [
      {
        msgKey: 'dashboard.card.deployables',
        textKey: 'dashboard.card.perInstance',
        count: getNumDeployables(item),
        border: 'right'
      },
      {
        msgKey: 'dashboard.card.deployments',
        textKey: 'dashboard.card.total',
        count: getNumDeployments(item)
      },
      {
        msgKey: 'dashboard.card.pending',
        textKey: 'dashboard.card.deployments',
        count: getNumPendingDeployments(item)
      },
      {
        msgKey: 'dashboard.card.inProgress',
        textKey: 'dashboard.card.deployments',
        count: getNumInProgressDeployments(item)
      },
      {
        msgKey: 'dashboard.card.failed',
        textKey: 'dashboard.card.deployments',
        count: getNumFailedDeployments(item),
        alert: true
      },
      {
        msgKey: 'dashboard.card.incidents',
        textKey: 'dashboard.card.total',
        count: incidentCount,
        border: 'left',
        alert: true
      }
    ]

    return (
      <div id="resource-overview" className="overview-content">
        {showAppDetails ? (
          <React.Fragment>
            <StructuredListModule
              title={staticResourceData.detailKeys.title}
              headerRows={staticResourceData.detailKeys.headerRows}
              rows={staticResourceData.detailKeys.rows}
              data={item}
            />
            {modulesRight.length > 0 && (
              <div className="overview-content-right">{modulesRight}</div>
            )}
            <div className="overview-content-bottom">{modulesBottom}</div>
          </React.Fragment>
        ) : !showExpandedTopology ? (
          <React.Fragment>
            <div className="overview-content-bottom overview-content-with-padding">
              <CountsCardModule data={countsCardData} />
            </div>
            {/* <div className="overview-content-bottom overview-content-with-padding">
              <ApplicationTopologyModule
                showExpandedTopology={showExpandedTopology}
                params={params}
                actions={actions}
              />
            </div> */}
            <div className="overview-content-bottom overview-content-with-padding-except-left">
              <ChannelsCardCarousel data={channelList} />
            </div>
            <div className="overview-content-bottom overview-content-with-padding">
              <InstancesTable
                staticResourceData={staticResourceData}
                resourceType={resourceType}
                getVisibleResources={getVisibleResources}
              />
            </div>
          </React.Fragment>
        ) : (
          <div className="overview-content-bottom overview-content-with-padding">
            <ApplicationTopologyModule
              showExpandedTopology={showExpandedTopology}
              params={params}
              actions={actions}
            />
          </div>
        )}
      </div>
    )
  }
)

ResourceOverview.contextTypes = {
  locale: PropTypes.string
}

ResourceOverview.propTypes = {
  item: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  modules: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  params: PropTypes.object,
  resourceType: PropTypes.object,
  staticResourceData: PropTypes.object
}

const mapStateToProps = (state, ownProps) => {
  const { resourceType, params } = ownProps
  const { HCMChannelList } = state
  const name = decodeURIComponent(params.name)
  const item = getSingleResourceItem(state, {
    storeRoot: resourceType.list,
    resourceType,
    name,
    predicate: resourceItemByName,
    namespace: params.namespace ? decodeURIComponent(params.namespace) : null
  })
  return { item, channelList: getChannelsList(HCMChannelList) }
}

export default withRouter(
  connect(mapStateToProps)(withLocale(ResourceOverview))
)
