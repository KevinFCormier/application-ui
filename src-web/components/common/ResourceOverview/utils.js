/*******************************************************************************
 * Licensed Materials - Property of IBM
 * 5737-E67
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
import R from 'ramda'
import {
  kindsToExcludeForDeployments,
  getResourcesStatusPerChannel
} from '../../ApplicationDeploymentPipeline/components/PipelineGrid/utils'
import { pullOutKindPerApplication } from '../../ApplicationDeploymentPipeline/utils'

// Method will take in an object and return back the channels mapped for display purposes
export const getChannelsList = channels => {
  if (channels && channels instanceof Array && channels.length > 0) {
    const mappedChannels = channels.map(channel => {
      // Will return back status as:
      // [0, 0, 0, 0, 0]
      // Pass, Fail, InProgress, Pending, Unidentified
      const status = getResourcesStatusPerChannel(channel, false)
      return {
        name: channel.name || '',
        counts: {
          completed: {
            total: status[0] + status[4]
          },
          inProgress: {
            total: status[2] + status[3]
          },
          failed: {
            total: status[1]
          }
        }
      }
    })
    return mappedChannels
  }
  return []
}

export const getNumClusters = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    const filtered = data.related.filter(elem => elem.kind === 'cluster')
    const numClusters = filtered.reduce(
      (acc, cur) => acc + (cur.items instanceof Array ? cur.items.length : 0),
      0
    )
    // number of clusters - 1 for local-cluster
    return numClusters > 0 ? numClusters - 1 : 0
  } else {
    return 0
  }
}

export const getNumDeployables = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    const filtered = data.related.filter(elem => elem.kind === 'deployable')
    return filtered.reduce(
      (acc, cur) => acc + (cur.items instanceof Array ? cur.items.length : 0),
      0
    )
  } else {
    return 0
  }
}

export const getNumDeployments = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    const filtered = data.related.filter(
      elem => !kindsToExcludeForDeployments.includes(elem.kind)
    )
    return filtered.reduce(
      (acc, cur) => acc + (cur.items instanceof Array ? cur.items.length : 0),
      0
    )
  } else {
    return 0
  }
}

export const getNumInProgressDeployments = data => {
  const status = getResourcesStatusPerChannel(data, false)
  return status[2] + status[3]
}

export const getNumFailedDeployments = data => {
  const status = getResourcesStatusPerChannel(data, false)
  return status[1]
}

export const getNumCompletedDeployments = data => {
  const status = getResourcesStatusPerChannel(data, false)
  return status[0] + status[4]
}

export const getNumPolicyViolations = data => {
  if (data && data.related instanceof Array && data.related.length > 0) {
    return 0
  } else {
    return 0
  }
}

// Given a current resource data INCLUDING its related resources
// we want to return all the channels.
// Channels are not returned inside related resources so we have to
// insepect each related subscription because it contains the channel
// -------------------------------------
// This method is not actually used but its got some good logic we could use later
// -------------------------------------
export const getResourceChannels = resourceData => {
  if (resourceData && resourceData.related) {
    const relatedData = resourceData.related
    // We want to pull only subscription data
    const filterToSubscriptions = x => x.kind == 'subscription'
    const subscriptionData = R.filter(filterToSubscriptions, relatedData)
    if (subscriptionData[0] && subscriptionData[0].items) {
      const channels = subscriptionData[0].items.map(sub => {
        return sub.channel
      })
      return channels
    }
    return []
  }
  return []
}

// Given an resource Object (item) and a list of resources Objects (appRelations)
// We want to go through the appRelations to see there is a resource UID that
// matches the item resource.
// If there is a match we want to return it and count it as common because
// it appears in both.
const findMatchingResource = (kind, item, appRelations) => {
  // Go through the relationship list
  const match = appRelations.map(resource => {
    const kindItems = resource.items || []
    const kindTwo = resource.kind || ''
    // go through each kinds items
    const matched = kindItems.map(itemTwo => {
      // If the UIDs match then we found a match
      if (kindTwo == kind && item._uid == itemTwo._uid) {
        return { kind: kind, items: [itemTwo] }
      }
    })
    // Filter out the undefined entries
    const filterOutUndefined = x => x != undefined
    const filteredData = R.filter(filterOutUndefined, matched)
    return filteredData
  })

  const filterOutEmpty = x => x.length > 0
  const filteredDataTwo = R.filter(filterOutEmpty, match)
  return filteredDataTwo
}

// We want to find the commone resources that between too objects
// EXAMPLE:
//  - We have one application with all of its related resources.
//  - We have a list of channels with all their relate resources.
//  - Given those TWO lists we want to find the common resources that are in
//    both the application related resources and the channels related resources.
//
// The reason is the channel resources wiill include ALL related resources in
// that particular channel which includes many different application resources
// if sepearate applications chose to put a subscription under that channel.
//
// -------------------------------------
// This method is not actually used but its got some good logic we could use later time
// -------------------------------------
export const getCommonResources = (channelData, appRelations) => {
  let channelFormat = []
  let commonResources = []
  // Go through Each Channel
  channelData.map(channel => {
    // That channels related data
    channel.related.map(resource => {
      const kindItems = resource.items || []
      const kind = resource.kind || ''
      // go through each kinds items
      kindItems.map(item => {
        const match = findMatchingResource(kind, item, appRelations)
        commonResources = commonResources.concat(match)
      })
    })
    // we want to tall up all the related resources for the given channel
    channelFormat = channelFormat.concat([
      { name: channel.name, items: commonResources }
    ])
  })
  // Filter out entries that are empty
  const filterOutUndefinedName = x => x.name != 'undefined'
  const filteredData = R.filter(filterOutUndefinedName, channelFormat)
  return filteredData
}

// We want to get the current Application if there is only ONE application.
// We check the secondaryHeader to see if there are two breadcrumbs which would
// mean that the user has selected one application to look into.
// We then extract the items from the application list and then pull out the application
// at location 0 which we know because there is only 1 application selected
// at this time.
export const getCurrentApplication = (HCMApplicationList, singleAppView) => {
  const currentApplication = R.pathOr([], ['items'])(HCMApplicationList)
  const currentApplicationItemZero =
    (singleAppView &&
      (currentApplication.length > 0 && currentApplication[0])) ||
    {}
  return currentApplicationItemZero
}

export const formatToChannel = (subscriptionList, bulkSubscription) => {
  let channelList = []
  subscriptionList instanceof Array &&
    subscriptionList[0] &&
    subscriptionList[0].items instanceof Array &&
    subscriptionList[0].items.map(subscription => {
      bulkSubscription &&
        bulkSubscription.items instanceof Array &&
        bulkSubscription.items.map(bulkSub => {
          const bulkSubChannelObj = pullOutKindPerApplication(
            bulkSub,
            'channel'
          )
          let bulkSubChannel = {}
          if (
            bulkSubChannelObj instanceof Array &&
            bulkSubChannelObj[0] &&
            bulkSubChannelObj[0].items instanceof Array &&
            bulkSubChannelObj[0].items[0]
          ) {
            bulkSubChannel = bulkSubChannelObj[0].items[0]
          }
          if (
            subscription.name == bulkSub.name &&
            subscription.namespace == bulkSub.namespace &&
            subscription.channel ==
              `${bulkSubChannel.namespace}/${bulkSubChannel.name}`
          ) {
            channelList = channelList.concat([
              {
                name: subscription.channel.split('/')[1],
                related: bulkSub.related
              }
            ])
          }
        })
    })
  return channelList
}

export const getSearchLinkForOneApplication = params => {
  if (params && params.name) {
    return `/multicloud/search?filters={"textsearch":"kind%3Aapplication%20name%3A${
      params.name
    }"}`
  }
  return ''
}

export const getSearchLinkForAllApplications = () => {
  return '/multicloud/search?filters={"textsearch":"kind%3Aapplication"}'
}

export const getSearchLinkForAllSubscriptions = () => {
  return '/multicloud/search?filters={"textsearch":"kind%3Asubscription"}'
}

export const getSearchLinkForAllChannels = () => {
  return '/multicloud/search?filters={"textsearch":"kind%3Achannel"}'
}
