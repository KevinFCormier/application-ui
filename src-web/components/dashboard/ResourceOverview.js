/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import lodash from 'lodash'
import PropTypes from 'prop-types'
import DashboardSection from '../common/DashboardSection'
import msgs from '../../../nls/platform.properties'
import { DashboardCard } from '../ComponentCard'
import {withRouter} from 'react-router-dom'

class ResourceOverview extends React.Component {
  render() {
    const { locale } = this.context
    const { cardItems = [] } = this.props
    return (
      <div id='component-summary'>
        <DashboardSection name={msgs.get('dashboard.section.resource-overview', locale)}>
          {cardItems.map(item =>
            item.name &&
            <DashboardCard
              critical={item.critical}
              warning={item.warning}
              healthy={item.healthy}
              type={item.type}
              title={msgs.get(`dashboard.card.${lodash.camelCase(item.name)}`, locale)}
              key={item.name}
              id={`card-${item.name.replace(' ', '-')}`}
            />)}
        </DashboardSection>
      </div>
    )
  }
}

DashboardSection.contextTypes = {
  locale: PropTypes.string
}

ResourceOverview.propTypes = {
  cardItems: PropTypes.arrayOf(PropTypes.shape({
    critical: PropTypes.number,
    healthy: PropTypes.number,
    warning: PropTypes.number,
    name: PropTypes.string,
    title: PropTypes.string,
    type: PropTypes.string,
  }))
}

export default withRouter(ResourceOverview)
