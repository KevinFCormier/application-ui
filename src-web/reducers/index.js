/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/*
 *
 * Things you should never do inside a reducer:
 *
 * - Mutate its arguments
 * - Perform side effects like API calls and routing transitions
 * - Call non-pure functions, e.g. Date.now() or Math.random()
 *
 * Reducers must be deterministic pure functions.  Given the same arguments, it should calculate the next state and return it.
 * No surprises. No side effects. No API calls. No mutations. Just a calculation.
 *
 * Selectors should sit along side reducers.
 */

import { createResourceReducer, resourceReducerFunction } from './common'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import lodash from 'lodash'

export { user, loggedIn } from './user'
export { secondaryHeader } from './common'

export { modal } from './modal'

//the exported function name must match the resourceType value
export const HCMApplicationList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_APPLICATIONS))
export const HCMClusterList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_CLUSTERS))
export const HCMComplianceList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_COMPLIANCES))
export const HCMDashboardList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_DASHBOARD))
export const HCMNodeList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_NODES))
export const HCMPersistentVolumeList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_PVS))
export const HCMPersistentVolumeClaimList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_PVS_CLAIM))
export const HCMPodList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_PODS))
export const HCMPolicyList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_POLICIES))
export const HCMReleaseList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_RELEASES))
export const HCMRepositoryList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_REPOSITORIES))

export { default as catalog } from './catalog'
export { topology } from './topology'
export { resourceFilters } from './filter'

function predicate(resourceType, action) {
  if (lodash.isEqual(resourceType, action.resourceType))
    return true
  return lodash.find(lodash.values(resourceType), type => {
    if (typeof type === 'string') {
      return type.indexOf(action.resourceType) > -1
    }
    return false
  })
}
