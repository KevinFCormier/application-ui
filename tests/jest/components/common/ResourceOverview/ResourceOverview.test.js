/*******************************************************************************
 * Licensed Materials - Property of Red Hat
 * Copyright (c) 2020 Red Hat, Inc.
 *******************************************************************************/

const React = require("../../../../../node_modules/react");

import ResourceOverview from "../../../../../src-web/components/common/ResourceOverview/";

import renderer from "react-test-renderer";
import * as reducers from "../../../../../src-web/reducers";

import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

const preloadedState = window.__PRELOADED_STATE__;
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const middleware = [thunkMiddleware];

const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
);

describe("ResourceOverview", () => {
  it("ResourceOverview renders spinner", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <ResourceOverview
              resourceType={resourceType}
              params={params}
              role={role}
            />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("ResourceOverview renders correctly with empty params", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <ResourceOverview
              resourceType={resourceType}
              params={emptyParams}
              role={role}
            />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const resourceType = {
  name: "QueryApplications",
  list: "QueryApplicationList"
};

const params = {
  name: "app1",
  namespace: "default"
};

const emptyParams = {
  name: undefined,
  namespace: undefined
};

const role = {
  role: "ClusterAdministrator",
  status: "DONE",
  type: "ROLE_RECEIVE_SUCCESS"
};
