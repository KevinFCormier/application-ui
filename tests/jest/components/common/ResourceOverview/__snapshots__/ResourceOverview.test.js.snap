// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceOverview ResourceOverview renders correctly with empty params 1`] = `
<div
  aria-live="assertive"
  className="bx--loading content-spinner"
>
  <svg
    className="bx--loading__svg"
    viewBox="-75 -75 150 150"
  >
    <circle
      cx="0"
      cy="0"
      r="37.5"
    />
  </svg>
</div>
`;

exports[`ResourceOverview ResourceOverview renders spinner 1`] = `
<div
  aria-live="assertive"
  className="bx--loading content-spinner"
>
  <svg
    className="bx--loading__svg"
    viewBox="-75 -75 150 150"
  >
    <circle
      cx="0"
      cy="0"
      r="37.5"
    />
  </svg>
</div>
`;
